docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -262,8 +262,151 @@ Every inter-canister call is guaranteed to receive a response, either from the c
+
+We proceed to discuss “journaling”, which is currently our recommended way of addressing the problem at hand. 
+
+### Journaling
+
+Journaling can be used for ensuring that tasks are completed correctly in an asynchronous context, where any instruction or async task can fail. Journaling is generally useful in any security critical application canister on the IC. The journaling concept we describe here is inspired and adapted from journaling in file systems.
+
+A journal is a chronological list of the records and is kept in the canister’s storage. It keeps track of tasks before they begin and when they are completed. Before each failable task, the journal records the intent to execute the task, and after the task, the journal records the result. Creating a record in the journal is called “journaling”.  For example, to make an unreliable async call to a ledger:
+
+1. Check the journal to ensure the transfer is not already in progress. If it is already in progress, go into recovery (see Recovery section below). Otherwise, journal the intent to call a ledger to transfer 1 token from A to B. The journaled intent should contain sufficient context to later identify what happened to the call.
+
+   - An “in progress” transfer would show in the journal as an entry containing intent to do the transfer without an entry containing the result of the transfer call.
+
+1. Call ledger to transfer 1 token from A to B.
+
+1. Journal the result of the transfer.
+
+   - On failure, record the error.
+
+   - On success, record success.
+
+1. Continue onto the next blocked task.
+
+   - “Blocked tasks” are those that require that step 3 has been completed before execution.
+
+   - A blocked task may depend on the success or failure recorded in step 3.
+
+   - Examples of blocked tasks:
+
+     - On failure, log the failure in a user-visible log, and if less than 5 failures have occurred, make a new transfer outcall with the same parameters.
+
+     - On success, update the internal accounting of assets to conform to the result of the transfer.
+
+   - Note that any independent task does not need to wait for any part of this flow.
+
+The critical property of the journal is that at any point, if there is a failure, the journal is sufficient to determine what the next safe step should be. If, after step 1 (journal the intent),
+there is a failure in step 2 or 3, and step 3 (record the result) has not been completed, then the application should complete step 3 by finding out what happened to the call in step 2.
```

Comment:
Yes, although it is tedious, I think there is a way to correctly check the status of a transaction.

We could make sure the ledger is past the 24 hour limit by resending the transaction and seeing if it says it is too old.

Then we should be able to scan the blocks backwards in time to check for our transaction. Asynchrony should not be a big problem because we can keep track of the block height as we scan. (It may be necessary to have checked the block height before sending the transaction, so we know when to stop scanning.) If we don't find the transaction then it must have failed.

Reply:
> Agree this seems tedious, also I guess we have no guarantees that a transaction will eventually make it to the archive, right? So one can never be sure that a transaction did _not_ happen?

@robin-kunzler I'm not sure I understood this point. Assuming that we did have a "has this transaction happened" endpoint on the ledger, our message ordering guarantees would ensure that once a "has the transaction happened" message makes it to the ledger, any prior "perform this transaction" message must also have made it to the ledger - or will never make it. So unless executing "perform this transaction" message is spread over multiple message handlers (because there are downstream calls involved), and unless the effects of the message are delayed, the second message would give you a definite answer.

Reply:
GoldDAO uses journaling to keep track of the process of acquiring a gold NFT and minting GLDT tokens. This happens in `notify_sale_nft_origyn`. AFAIK GoldDAO also requires manual recovery on trap. However, it may be that automated recovery can be implemented too much trouble. It depends on the Origyn NFT interface for determining the status of one's attempt to acquire an NFT. I would expect it's easier than checking the status of a ledger transaction. At least GoldDAO can determine if they own the NFT.

We also know of another project that does journaling and also uses manually recovery on failures. There is special functionality for principals with admin privileges to refund funds. In this project the journal is for dealing with ledger call failures, so it would tedious to resolve automatically. (Note that this is not yet open sourced so I won't reference it explicitly. We could consider adding the link later when the project is published.)

I agree we should at least recommend journaling with manual recovery. Early-stage projects that we've looked at that contain journaling appear to do only manual recovery. However, as in the case of GoldDAO, it can be practical to do automated recovery as the project matures.

Note that there is an index canister for SNSs. I would need to check but there might be one for the ICP ledger too. The index canister indexes ledger transactions for easier searching.